packages/vite/src/node/plugins/asset.ts

```diff
@@ -431,8 +431,10 @@ function svgToDataURL(content: Buffer): string {
       stringContent
         .trim()
         .replaceAll(/>\s+</g, '><')
-        .replaceAll('"', "'")
         .replaceAll('%', '%25')
+        .replaceAll(/"([^"']*'[^"]*)"/g, `%27$1%27`)
+        .replaceAll(/(?<=(%27.*))'(?!(%27))/g, `%23`)
+        .replaceAll('"', "'")
         .replaceAll('#', '%23')
         .replaceAll('<', '%3c')
         .replaceAll('>', '%3e')
```

Comment:
I have another idea. How about changing the above to something just like this?

```js
'data:image/svg+xml,' + encodeURI(stringContent)
```

After all it is doing URI encoding. So, I think it is better to use the given function.

Reply:
However, it does break `playground/assets/__tests__/assets.spec.ts > svg fragments > from js import` and escaping of `data-test="vite's the world's best"`.

Reply:
I think the idea is that SVGs don't need a full encode since it already works without, and would otherwise that make the string even longer. So we only encode those for HTML attribute usage (`src="data:image..."`) and CSS (`url(...)`).